File: husky_mpc/husky_mpc/nmpc.py
from time import time
import time as tm
import casadi as ca
import numpy as np
from numpy.linalg import norm



class NMPC:
    def __init__(self, prediction_states, prediction_controls, Q, R, n_states, n_controls, N, max_linear_acc, max_angular_acc):
        self.prediction_states = prediction_states
        self.prediction_controls = prediction_controls
        self.Q = Q
        self.R = R
		
        self.n_states = n_states
        self.n_controls = n_controls
        self.N = N  # prediction horizons

        self.max_linear_acc = max_linear_acc
        self.max_angular_acc = max_angular_acc

        self.lbx = ca.DM.zeros((self.n_states*(self.N+1) + self.n_controls*self.N, 1))
        self.ubx = ca.DM.zeros((self.n_states*(self.N+1) + self.n_controls*self.N, 1))


    def compute_cost(self, P, time_interval, dynamics_function):
        cost_fn = 0  # cost function

        g = self.prediction_states[:, 0] - P[: self.n_states]
        # runge kutta
        for k in range(self.N):
            current_state = self.prediction_states[:, k]
            # control = U[:, k] - K_matrix@current_state
            control = self.prediction_controls[:, k]
    
            next_state = self.prediction_states[:, k+1]
            distance_difference = next_state - P[(k+1)*self.n_states:(k+2)*self.n_states, :]
            cost_fn = cost_fn \
                + distance_difference.T @ self.Q @ distance_difference \
                + control.T @ self.R @ control
    
            k1 = dynamics_function(current_state) @ control
            k2 = dynamics_function(current_state + time_interval/2*k1) @ control
            k3 = dynamics_function(current_state + time_interval/2*k2) @ control
            k4 = dynamics_function(current_state + time_interval * k3) @ control
            next_state_RK4 = current_state + (time_interval / 6) * (k1 + 2 * k2 + 2 * k3 + k4)
            g = ca.vertcat(g, next_state - next_state_RK4)

        return cost_fn, g 


    # obstacle avoidance
    def generate_obstacle_avoidance_constraints(self, obstacle_list):
        distance = []
        mask = ca.DM([[1, 0, 0], 
                  [0, 1, 0]]) 
        
        for i in range(len(obstacle_list)):
            obstacle = obstacle_list[i]
            #state_obstacle = ca.DM([obstacle.x, obstacle.y, 0]) 
            state_obstacle = ca.DM([obstacle.x, obstacle.y]) 

            for k in range(self.N):
                current_state = self.prediction_states[:, k]
                # we only consider x and y, not including theta
                distance = ca.vertcat(distance, ca.norm_2(mask@current_state - state_obstacle))
        return distance
    

    # control constraints
    def generate_acceleration_contraints(self, time_interval):
        linear_acc_contraints = []
        angular_acc_contraints = []

        for k in range(self.N-1):
            # a = dv/dt
            linear_acceleration = (self.prediction_controls[0, k+1] - self.prediction_controls[0, k])/time_interval
            angular_acceleration = (self.prediction_controls[1, k+1] - self.prediction_controls[1, k])/time_interval

            linear_acc_contraints = ca.vertcat(linear_acc_contraints, linear_acceleration)
            angular_acc_contraints = ca.vertcat(angular_acc_contraints, angular_acceleration)
        
        acceleration_contraints = ca.vertcat(linear_acc_contraints, angular_acc_contraints)
        return acceleration_contraints


    def define_problem(self, P, dt, dynamics_function, obstacle_list):
        total_cost_nmpc, g_nmpc = self.compute_cost(P, dt, dynamics_function)
        distance_to_obstacle = self.generate_obstacle_avoidance_constraints(obstacle_list)

        # linear and angular acceleration contraints
        #acc_nmpc = self.generate_test_contraints()
        acc_nmpc = self.generate_acceleration_contraints(dt)

        OPT_variables_nmpc = ca.vertcat(
            self.prediction_states.reshape((-1, 1)),   # Example: 3x11 ---> 33x1 where 3=states, 11=N+1
            self.prediction_controls.reshape((-1, 1))
        )

        nlp_prob_nmpc = {
            'f': total_cost_nmpc,
            'x': OPT_variables_nmpc,
            'g': ca.vertcat(distance_to_obstacle, g_nmpc, acc_nmpc),
            'p': P
        }

        return nlp_prob_nmpc
    

	# GENERATE x, y and theta CONSTRAINTS FOR NMPC
    # TODO: add map parameters as input arguments 
    def generate_state_constraints(self, current_state, noise, safety_radius, lower_x_constraint, upper_x_constraint, lower_y_constraint, upper_y_constraint, 
                                   min_linear_speed, max_linear_speed, min_angular_speed, max_angular_speed, obstacle_list):
        noised_state = current_state + noise

        # state lower bounds
        self.lbx[0: self.n_states*(self.N+1): self.n_states] = lower_x_constraint

        # look up lower y constraints
        for i in range(self.N+1):
            self.lbx[1+ self.n_states*i] = lower_y_constraint
        self.lbx[2: self.n_states*(self.N+1): self.n_states] = -ca.pi


        # state upper bounds
        self.ubx[0: self.n_states*(self.N+1): self.n_states] = upper_x_constraint
        
        # look up upper y constraints
        for i in range(self.N+1):
            self.ubx[1+ self.n_states*i] = upper_y_constraint
        self.ubx[2: self.n_states*(self.N+1): self.n_states] = ca.pi

        # lower and upper bounds for u
        for k in range(self.N):
            self.lbx[self.n_states*(self.N+1) + self.n_controls*k: self.n_states*(self.N+1) + self.n_controls*(k+1)] = ca.DM([min_linear_speed, min_angular_speed])
            self.ubx[self.n_states*(self.N+1) + self.n_controls*k: self.n_states*(self.N+1) + self.n_controls*(k+1)] = ca.DM([max_linear_speed, max_angular_speed])
        
        args_nmpc = {
            'lbg': ca.vertcat(safety_radius*ca.DM.ones((len(obstacle_list) * self.N, 1)), 
                              ca.DM.zeros((self.n_states*(self.N+1), 1)),
                              -1*self.max_linear_acc*ca.DM.ones(self.N-1, 1),
                              -1*self.max_angular_acc*ca.DM.ones(self.N-1, 1)
                            ),  # constraints lower bound
            'ubg': ca.vertcat(400*ca.DM.ones((len(obstacle_list) * self.N, 1)), 
                              ca.DM.zeros((self.n_states*(self.N+1), 1)),
                              self.max_linear_acc*ca.DM.ones(self.N-1, 1),
                              self.max_angular_acc*ca.DM.ones(self.N-1, 1)
                            ),  # constraints upper bound
            'lbx': self.lbx,
            'ubx': self.ubx
        }
        args_nmpc['p'] = noised_state

        return args_nmpc

File: husky_mpc_cpp/src/Nmpc.cpp
#include "husky_mpc_cpp/Nmpc.hpp"
#include <iostream>
#include <algorithm>    // std::min

NMPC::NMPC(const casadi::SX& prediction_states, const casadi::SX& prediction_controls, const casadi::DM& Q, const casadi::DM& R, 
            int n_states, int n_controls, int N, double max_linear_acc, double max_angular_acc) : 
        prediction_states_(prediction_states), 
        prediction_controls_(prediction_controls),
        Q_(Q), 
        R_(R), 
        n_states_(n_states), 
        n_controls_(n_controls), 
        N_(N), 
        max_linear_acc_(max_linear_acc), 
        max_angular_acc_(max_angular_acc)
{
    lbx_ = casadi::DM::zeros((n_states_*(N_+1) + n_controls_*N_), 1);
    ubx_ = casadi::DM::zeros((n_states_*(N_+1) + n_controls_*N_), 1);
}



// Cost based on states and controls
std::tuple<casadi::SX, casadi::SX> NMPC::compute_cost(const casadi::SX& P, double time_interval, const casadi::Function& dynamics_function)
{
    casadi::SX cost_fn = 0;
    casadi::SX g = prediction_states_(casadi::Slice(), 0) - P(casadi::Slice(0, n_states_)); // Difference between actual and predicted states

    // Runge Kutta-4 Estimation
    for (int k = 0; k < N_; ++k)
    {
        casadi::SX current_state = prediction_states_(casadi::Slice(), k);
        casadi::SX control = prediction_controls_(casadi::Slice(), k);
        casadi::SX next_state = prediction_states_(casadi::Slice(), k+1); // Predicted next state
        casadi::SX distance_difference = next_state - P(casadi::Slice((k+1)*n_states_, (k+2)*n_states_), casadi::Slice());
        
        casadi::SX state_cost = mtimes(mtimes(distance_difference.T(), casadi::SX(Q_)), distance_difference);
        casadi::SX control_cost = mtimes(mtimes(control.T(), casadi::SX(R_)), control);
        cost_fn = cost_fn + state_cost +  control_cost;
        casadi::SX k1 = mtimes(dynamics_function(current_state)[0], control);
        casadi::SX k2 = mtimes(dynamics_function(current_state + time_interval/2*k1)[0], control);
        casadi::SX k3 = mtimes(dynamics_function(current_state + time_interval/2*k2)[0], control);
        casadi::SX k4 = mtimes(dynamics_function(current_state + time_interval * k3)[0], control);
        casadi::SX next_state_RK4 = current_state + (time_interval / 6) * (k1 + 2 * k2 + 2 * k3 + k4);
        g = vertcat(g, next_state - next_state_RK4); // Difference between next state of optimal trajectory and next state approximation
    }
    return std::tuple<casadi::SX, casadi::SX>(cost_fn, g); 
}



// Obstacle avoidance constraints
casadi::SX NMPC::generate_obstacle_avoidance_constraints(const std::vector<Obstacle>& obstacle_list)
{
    casadi::SX distance;
    auto mask = casadi::DM({{1, 0, 0}, {0, 1, 0}}); // Mask to extract x and y coordinates from state vector
    for (auto obstacle : obstacle_list)
    {
        casadi::DM state_obstacle({obstacle.x_, obstacle.y_});
        for (int k = 0; k < N_; ++k)
        {
            auto current_state = prediction_states_(casadi::Slice(), k);
            distance = vertcat(casadi::SX(distance), norm_2(mtimes(casadi::SX(mask), current_state) - casadi::SX(state_obstacle))); // Append current obstacle distance to distance list
        }
    }
    return distance;
}

// Acceleration constraints
casadi::SX NMPC::generate_acceleration_constraints(double time_interval)
{
    casadi::SX linear_acc_contraints;
    casadi::SX angular_acc_contraints;
    for (int k = 0; k < N_-1; ++k)
    {
        casadi::SX linear_acceleration = (prediction_controls_(0, k+1) - prediction_controls_(0, k)) / time_interval;
        casadi::SX angular_acceleration = (prediction_controls_(1, k+1) - prediction_controls_(1, k)) / time_interval;
        linear_acc_contraints = vertcat(linear_acc_contraints, linear_acceleration);
        angular_acc_contraints = vertcat(angular_acc_contraints, angular_acceleration);
    }
    return vertcat(linear_acc_contraints, angular_acc_contraints);
}


std::map<std::string, casadi::DM> NMPC::generate_state_constraints(double lower_x_constraint, double upper_x_constraint, 
                                                                   double lower_y_constraint, double upper_y_constraint, 
                                                                   double current_linear_v, double current_angular_v,
                                                                   double max_linear_speed, double max_angular_speed, 
                                                                   double time_interval, const std::vector<Obstacle>& obstacle_list)
{
    // STATE LOWER BOUNDS
    lbx_(casadi::Slice(0, n_states_*(N_+1), n_states_)) = lower_x_constraint;

    //look up lower y constraints
    for (int i = 0; i < N_+1; ++i)
    {
        lbx_(1 + n_states_*i) = lower_y_constraint;
    }
    //constraint for theta
    lbx_(casadi::Slice(2, n_states_*(N_+1), n_states_)) = -casadi::pi;


    // STATE UPPER BOUNDS
    ubx_(casadi::Slice(0, n_states_*(N_+1), n_states_)) = upper_x_constraint;

    //look up upper y constraints
    for (int i = 0; i < N_+1; ++i)
    {
        ubx_(1 + n_states_*i) = upper_y_constraint;
    }
    ubx_(casadi::Slice(2, n_states_*(N_+1), n_states_)) = casadi::pi;



    // LOWER and UPPER BOUNDS for CONTROL u
    for (int k = 0; k < N_; ++k)
    {
        if (k == 0){
            // first step in prediction horizon
            double change_of_linear_speed = max_linear_acc_*time_interval;
            double change_of_angular_speed = max_angular_acc_*time_interval;

            lbx_(n_states_*(N_+1) + n_controls_*k) = std::max(-1*max_linear_speed, -1*change_of_linear_speed+current_linear_v);
            lbx_(n_states_*(N_+1) + n_controls_*k + 1) = std::max(-1*max_angular_speed, -1*change_of_angular_speed+current_angular_v);
            
            ubx_(n_states_*(N_+1) + n_controls_*k) = std::min(max_linear_speed, change_of_linear_speed+current_linear_v);
            ubx_(n_states_*(N_+1) + n_controls_*k + 1) = std::min(max_angular_speed, change_of_angular_speed+current_angular_v);
        } else {
            lbx_(n_states_*(N_+1) + n_controls_*k) = -1 * max_linear_speed;
            lbx_(n_states_*(N_+1) + n_controls_*k + 1) = -1 * max_angular_speed;
            ubx_(n_states_*(N_+1) + n_controls_*k) = max_linear_speed;
            ubx_(n_states_*(N_+1) + n_controls_*k + 1) = max_angular_speed;
        }
        
    }


    std::map<std::string, casadi::DM> args_nmpc = {
        // TODO: update constraints
        {"lbg", vertcat(4*casadi::DM::ones(obstacle_list.size() * N_, 1), 
                          casadi::DM::zeros(n_states_*(N_+1), 1),
                          -1*max_linear_acc_*casadi::DM::ones(N_-1, 1),
                          -1*max_angular_acc_*casadi::DM::ones(N_-1, 1)
                        )},  // constraints lower bound
        {"ubg", vertcat(400*casadi::DM::ones(obstacle_list.size() * N_, 1), 
                          casadi::DM::zeros(n_states_*(N_+1), 1),
                          max_linear_acc_*casadi::DM::ones(N_-1, 1),
                          max_angular_acc_*casadi::DM::ones(N_-1, 1)
                        )},  // constraints upper bound
        {"lbx", lbx_},
        {"ubx", ubx_}
    };

    return args_nmpc;
}



// Define problem
std::map<std::string, casadi::SX> NMPC::define_problem(const casadi::SX& P, double dt, const casadi::Function& dynamics_function, const std::vector<Obstacle>& obstacle_list)
{
    // Call compute_cost
    auto cost_tuple = compute_cost(P, dt, dynamics_function); // Error here
    casadi::SX total_cost_nmpc = std::get<0>(cost_tuple);
    casadi::SX g_nmpc = std::get<1>(cost_tuple);


    // Generate obstacle avoidance constraints
    casadi::SX distance_to_obstacle = generate_obstacle_avoidance_constraints(obstacle_list);

    // Generate acceleration constraints
    casadi::SX acc_nmpc = generate_acceleration_constraints(dt);
    // Concatenate prediction_states and prediction_controls
    casadi::SX OPT_variables_nmpc = vertcat(reshape(prediction_states_, -1, 1), reshape(prediction_controls_, -1, 1));
    std::map<std::string, casadi::SX> nlp_prob_nmpc = {
        {"f", total_cost_nmpc},
        {"x", OPT_variables_nmpc},
        {"g", vertcat(distance_to_obstacle, g_nmpc, acc_nmpc)},
        {"p", P}
    };

    return nlp_prob_nmpc;
}
